### PR TITLE
Django backend exercise

### PR DESCRIPTION
**Pull request by betuca-tk**
Create a CRUD API with Django and DRF that allows you to CRUD recipes and add/delete ingredients to it. Exercise details can be found [here](https://www.notion.so/travelperk/Description-of-the-Django-Exercise-49da5411fb7149ed844d70baebf93efc).

### CODE REVIEW

File: app/recipes_book/tests/test_recipe_api.py [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+        self.assertEqual(response.status_code, 204)
+
+        with self.assertRaises(Recipe.DoesNotExist):
+            Recipe.objects.get(name=self.recipe_one.name)
+
+        with self.assertRaises(Ingredient.DoesNotExist):
+            Ingredient.objects.get(name=self.ingredient_two.name)
+
+        with self.assertRaises(Ingredient.DoesNotExist):
+            Ingredient.objects.get(name=self.ingredient_two.name)
```
Comment (oriol-travelperk) [inline app/recipes_book/tests/test_recipe_api.py]: Should this be ingredient `one` and `two`?

File: app/recipes_book/views.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+            filter_name = request.GET.get("name")
+            recipes = self._load_recipes_by_name(filter_name)
+            return JsonResponse(recipes, safe=False)
+        else:
+            recipes = self._load_all_recipes()
+            return JsonResponse(recipes, safe=False)
+
+    def _load_by_id(self, recipe_id):
+        try:
+            recipe = models.Recipe.objects.prefetch_related('ingredients').get(id=recipe_id)
```
Comment (oriol-travelperk) [inline app/recipes_book/views.py]: `prefetch_related` nice! 🚀

File: app/recipes_book/views.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    def _load_recipes_by_name(self, name):
+        recipes = models.Recipe.objects.filter(name__icontains=name)
+        recipes_result = [recipe.to_dict() for recipe in recipes]
+        return recipes_result
+
+    def post(self, request):
+        data = json.loads(request.body)
+        recipe = models.Recipe.objects.create(
+            name=data["name"], description=data["description"]
+        )
```
Comment (oriol-travelperk) [inline app/recipes_book/views.py:45]: Should there be some validation here? Like no empty name or description?

File: app/recipes_book/views.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+    def patch(self, request, recipe_id):
+        try:
+            recipe = models.Recipe.objects.get(id=recipe_id)
+            data = json.loads(request.body)
+            recipe.name = data["name"] if "name" in data else recipe.name
+            recipe.description = (
+                data["description"] if "description" in data else recipe.description
+            )
+            recipe.save()
```
Comment (oriol-travelperk) [inline app/recipes_book/views.py:68]: Not sure but maybe this can be simplified with `recipe.update(...)`

**Review by oriol-travelperk — COMMENTED** [review]
Looks very good 👍

File: app/recipes_book/views.py (hunk 2 above)
Comment (betuca-tk) [reply app/recipes_book/views.py:22]: it does. Will fix that.

File: app/recipes_book/tests/test_recipe_api.py (hunk 1 above)
Comment (betuca-tk) [reply app/recipes_book/tests/test_recipe_api.py]: it does. Will fix that.

File: app/recipes_book/views.py (hunk 3 above)
Comment (betuca-tk) [reply app/recipes_book/views.py:45]: Makes sense. I will create a test to cover this too.

File: app/recipes_book/views.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+                recipe.ingredients.all().delete()
+                self._create_ingridients(recipe, data)
+            return JsonResponse(recipe.to_dict(), status=200)
+        except models.Recipe.DoesNotExist:
+            raise Http404("Recipe not found")
+
+    def _create_ingridients(self, recipe, data):
+        for ingredient in data["ingredients"]:
+            ingredient = models.Ingredient.objects.create(
+                name=ingredient["name"], recipe=recipe
```
Comment (devashishsharma2302) [inline app/recipes_book/views.py]: To make this more efficient, we can use the `bulk_create` [method from django](https://docs.djangoproject.com/en/5.0/ref/models/querysets/#bulk-create) here. 🙂  

By using the bulk_create, we can create all those objects in a single database query.

File: app/recipes_book/views.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+            return JsonResponse(recipes, safe=False)
+
+    def _load_by_id(self, recipe_id):
+        try:
+            recipe = models.Recipe.objects.prefetch_related("ingredients").get(
+                id=recipe_id
+            )
+            return recipe.to_dict()
+        except models.Recipe.DoesNotExist:
+            raise Http404("Recipe not found")
```
Comment (devashishsharma2302) [inline app/recipes_book/views.py]: checkout https://docs.djangoproject.com/en/5.0/topics/http/shortcuts/#get-object-or-404 🙂

**Review by devashishsharma2302 — APPROVED** [review]
Looks very good! 🙂 

Just for reference, there is also an alternate way to create the views using Djnago rest framework. by using ModelViewSets and serialisers.  More info here: https://www.django-rest-framework.org/api-guide/viewsets/#modelviewset

File: app/recipes_book/views.py (hunk 5 above)
Comment (betuca-tk) [reply app/recipes_book/views.py]: Sounds good! I will take a loot and try that!

File: app/recipes_book/views.py (hunk 6 above)
Comment (betuca-tk) [reply app/recipes_book/views.py]: Tried and it really looks better as you suggested. 👍

**Pull request merged**